File: asr/android/sdk-dingqiao/src/main/java/com/amphion/dingqiao/diarization/DiarizationCommitClock.kt
package com.amphion.dingqiao.diarization

internal data class DiarizationCommitBoundary(
    val beginTime: Int,
    val endTime: Int,
    val evidenceEndTime: Int,
)

internal class DiarizationCommitClock {
    private var plannedThrough = 0
    private var committedThrough = 0
    private val pending = java.util.ArrayDeque<DiarizationCommitBoundary>()

    fun observeEndpoint(endTime: Int) {
        if (endTime < plannedThrough + 120_000) return
        pending.add(DiarizationCommitBoundary(plannedThrough, endTime,
            ((endTime + 1500 + 2499) / 2500) * 2500))
        plannedThrough = endTime
    }

    fun takeReady(inferenceEndTime: Int): DiarizationCommitBoundary? {
        val next = pending.peekFirst() ?: return null
        if (inferenceEndTime < next.evidenceEndTime) return null
        pending.removeFirst()
        committedThrough = next.endTime
        return next
    }

    fun beginTime(): Int = committedThrough
}

File: asr/harmony/sdk-dingqiao/src/main/ets/com/amphion/dingqiao/diarization/DiarizationCommitClock.ts
export interface DiarizationCommitBoundary {
  beginTime: number;
  endTime: number;
  evidenceEndTime: number;
}

/** Audio-time decisions are independent of which inference/callback queue happens to run first. */
export class DiarizationCommitClock {
  private plannedThrough: number = 0;
  private committedThrough: number = 0;
  private processedThrough: number = 0;
  private endpoints: number[] = [];
  private readonly pending: DiarizationCommitBoundary[] = [];

  observeEndpoint(endTime: number): void {
    if (endTime > this.plannedThrough &&
      endTime > (this.endpoints[this.endpoints.length - 1] ?? 0)) this.endpoints.push(endTime);
  }

  /** Called only after ASR has returned all callbacks for the processed PCM. */
  observeProcessedAudio(endTime: number): boolean {
    this.processedThrough = Math.max(this.processedThrough, endTime);
    let queued = false;
    while (this.processedThrough >= this.plannedThrough + 120000) {
      const deadline = this.plannedThrough + 120000;
      let boundary: number | undefined;
      for (const endpoint of this.endpoints) {
        if (endpoint <= deadline) boundary = endpoint;
        else {
          // No completed sentence before the deadline: wait for the first one after it.
          if (boundary === undefined) boundary = endpoint;
          break;
        }
      }
      if (boundary === undefined || boundary > this.processedThrough) break;
      this.pending.push({ beginTime: this.plannedThrough, endTime: boundary,
        evidenceEndTime: Math.ceil((Math.max(deadline, boundary) + 1500) / 2500) * 2500 });
      this.plannedThrough = boundary;
      this.endpoints = this.endpoints.filter(endpoint => endpoint > boundary!);
      queued = true;
    }
    return queued;
  }

  takeReady(inferenceEndTime: number): DiarizationCommitBoundary | undefined {
    const next = this.pending[0];
    if (next === undefined || inferenceEndTime < next.evidenceEndTime) return undefined;
    this.pending.shift();
    this.committedThrough = next.endTime;
    return next;
  }

  beginTime(): number { return this.committedThrough; }
}
